Desktop: Added support for templates

CalebJohn opened the pull request:
I decided to write the first pass of this exactly the same as userstyle.css (i.e. template files go in the config folder). 
The current behaviour is that the user can place any number of markdown (.md) files in `.config/joplin-desktop/templates` and joplin will pick them up. If the user has placed any then whenever they create a new note or todo they will face a prompt to select a template. Leaving the input field blank (or cancelling) will fall back on an empty note. Selecting a template file will automatically populate the body with the contents of that file.

![image](https://user-images.githubusercontent.com/2179547/59325283-ae17e100-8c9f-11e9-9ec9-f370fd25960a.png)

![image](https://user-images.githubusercontent.com/2179547/59325217-6db86300-8c9f-11e9-8c15-756c6dc60dfb.png)

While I would like have the [syncing functionality](https://discourse.joplinapp.org/t/greyed-checkboxes/2499/26) in before adding templates. I felt that these should be 2 separate things and templates are a bit easier.

TODO
- [x] update readme
- [x] add placeholders, e.g automatically inserting date etc.
- [x] add some example templates

fixes #1585

laurent22 commented:
Thanks for the PR @CalebJohn, that looks good although indeed we don't have sync if we use files in a directory. 

That makes me think, I wonder if we should just create a new generic sync item type that would just be a title and a body. That item could be used for any text content we want to sync but that's not a note. Then it can be used for config, template, even scripts, etc. without affecting user notes and notebooks. Basically something that can be built on, while taking advantage of the sync and cross platform functionalities.

I'll check how easy it is to add this feature, but if you can think of some drawback let me know.

CalebJohn commented:
How would the user edit these? They would connect them to joplin through the config screen?

laurent22 commented:
Yes there would have to be some kind of editor and some way to list the items.

laurent22 commented:
By the way, for the placeholder, dates, etc., do you have a particular format in mind? Personally I always like the [Mustache](https://mustache.github.io/mustache.5.html) syntax as it's simple and supported by many languages including JavaScript.

CalebJohn commented:
I like the {{mustache}} syntax, for now I'm planning to do pretty simple templates, but it would be cool to expose more of the templating features to the user, for now, I plan to just have date and time, And if I (or anyone else) thinks of something else that's useful, I can add it to this pull.

CalebJohn commented:
I've added mustache templating (just for date and time) to the new note templates and updated the readme. That concludes all the features that I was planning initially, but I want to wait to see if we get any responses on [the forum post I made](https://discourse.joplinapp.org/t/jopline-templates/2667/3) and I want to do a bit of searching for some example templates that might be nice to link to or host.

laurent22 commented on ElectronClient/app/app.js:
```diff
@@ -1037,6 +1069,13 @@ class Application extends BaseApplication {
+		const templates = await this.loadTemplates(Setting.value('profileDir') + '/templates');
+
+		this.store().dispatch({
+			type: 'LOAD_TEMPLATES',
```
The action type format is "TYPE_VERB" (type is singular), for example "NOTE_SELECT". So in your case it would be "TEMPLATE_UPDATE_ALL" for consistency with the other actions.

CalebJohn replied:
Thanks for the direction I've changed it to TEMPLATE_UPDATE_ALL

laurent22 commented:
Your implementation of custom date/time format is pretty good, as the result is a valid Mustache template (unlike my suggestion which was kind of a hack).

As I understand, the template dialog will show up every time you create a new note, is that right? I think we should avoid this extra step if the user doesn't need it. Ideally, I think there should be three menu items to deal with templates:

- "Create note from template" (what you have now)
- "Insert template into note" (basically it would let the user select a template, and insert it at the cursor position)
- "Reveal template directory", which would open the template directory so that you can quickly open it without looking for it on your computer.

I guess these three menu items could be under a "Template" menu under "File".

The way I see, that would be the minimum we need to make the feature more user friendly, what do you think?

As for my suggestion earlier of creating a special sync item, let's leave it out for now. As a first step, we can get the feature pretty much as you've implemented it. Later if needed we can improve it with these special sync items.

CalebJohn commented:
I had initially started that way, but a few day's ago changed it so that a prompt would appear in the note editor/viewer area. This prompt is only visible when the note body is empty.

![image](https://user-images.githubusercontent.com/2179547/59823002-d203a400-92e9-11e9-92ec-9e13c505b876.png)

Currently I'm trying to make that prompt a little less ugly...

I like your suggestions for menu items, especially in conjuction with the prompt I'm planning. It's smart to go for the extra usability.

laurent22 commented:
I was thinking about this again and I wonder if all the template code should be moved to a separate service. That will encapsulate the template functionalities and allow us to change the loading method and rendering more easily.

Later I'd like to use this templating system to allow loading snippets from the Joplin Web Clipper, and then inserting them into pages (at some point I was looking for such a snippet extension and I thought it would be even better if all the snippets were in Joplin, and your system would allow that).

So if it's not too much to ask, is there any chance you could move this template code for example to lib/services/TemplateService.js? Then it would just have your `loadTemplates(path)` method and maybe you could also move the Mustache code and template variables to it. You could for example add a method `renderTemplate(body)` that returns the rendered template. The service needs to be a singleton and extend BaseService (see for example MigrationService to see how it's done). If you have any questions, please let me know.

CalebJohn commented:
I've done something similar already (I moved all template code into TemplateUtils.js). I'll push up my current wip right away for you to take a look. I'll also start looking into adding it as a service, I'll have more time for this tonight.

CalebJohn commented:
I don't quite understand the benefit of using a service rather than just a collection of util functions (what I have started in the latest commit). Is it for forward compatibility? Or is it for the logging maybe?

laurent22 commented on ElectronClient/app/app.js:
```diff
@@ -551,6 +571,13 @@ class Application extends BaseApplication {
 				visible: shim.isMac() ? false : true
+			}, {
+				label: _('Templates'),
+				visible: shim.isMac() ? false : true,
```
Any reason why this is not visible on macOS?

CalebJohn replied:
I was just following the surrounding code, I add the templates in `rootMenuFileMacOs` for Mac os just below this code

laurent22 replied:
This is fine then, I missed that in the context of the diff.

laurent22 commented:
Utility functions as you've done is fine as well and indeed for now we don't need a service. If we do at some point, it shouldn't be too hard to refactor anyway.

CalebJohn commented:
Phew I think this is finally done. Was definitely a bigger change than I anticipated. Please let me know if there is anything that can be changed. 

That said, I don't really like the look of the prompt that comes up on new notes, if anyone has suggestions to make it look better please let me know!

laurent22 commented:
@CalebJohn, sorry for the lack of feedback on this issue recently. I'm planning to test it properly at some point and get this merged, probably over the weekend. In the meantime thanks for your patience!

CalebJohn commented:
No worries, I understand that this is a rather large, non-critical pull. Take your time.

laurent22 commented on ElectronClient/app/app.js:
```diff
@@ -500,6 +503,45 @@ class Application extends BaseApplication {
 
+		templateItems.push({
+			label: _('Create note from template'),
+			visible: fs.pathExistsSync(Setting.value('profileDir') + '/templates'),
```
I think this whole block should not be executed (templateItems remains an empty array) if the templates folder does not exist. And we should simply hide or show the root template inside of each individual item. Also please use `shim.fsDriver().exists` (async) to test for the existence of a file.

laurent22 replied:
Edit: Sorry I missed that "Open template directory" should always be visible. I guess if you could just cache the value of `shim.fsDriver().exists(Setting.value('profileDir') + '/templates')` that would be enough. That way the check won't be repeated for each menu item.

CalebJohn replied:
Done! I had to make the entire function async.

laurent22 commented on ElectronClient/app/app.js:
```diff
@@ -1037,6 +1091,13 @@ class Application extends BaseApplication {
 			css: cssString
 		});
 
+		const templates = await TemplateUtils.loadTemplates(Setting.value('profileDir') + '/templates');
```
Could you create a "templateDir" constant for this (using `Setting.setConstant(...`)?

CalebJohn replied:
Done!

laurent22 commented on ElectronClient/app/gui/NoteText.jsx:
```diff
@@ -454,14 +455,18 @@ class NoteTextComponent extends React.Component {
 		if (props.newNote) {
-			note = Object.assign({}, props.newNote);
+			// assign new note and prevent body from being null
+			note = Object.assign(props.newNote, {body: ''});
```
I don't quite understand the reason for this change. I believe it will directly modify the component prop, which should not be done. I think it should be `Object.assign({}, props.newNote, {body: ''});`

CalebJohn replied:
The reason for that is because when using a template we assume that the note has a body (this can be changed), which I think is reasonable because it doesn't make sense for the note body to be null.
The reason for formatting the syntax that way is simply my own javascript ineptitude.

laurent22 commented on ElectronClient/app/gui/NoteText.jsx:
```diff
@@ -1021,6 +1026,8 @@ class NoteTextComponent extends React.Component {
 			} else if (command.name === 'textCode') {
 				fn = this.commandTextCode;
+			} else if (command.name === 'insertTemplate') {
+				fn = () => { return this.commandTemplate(command.value); };
```
Could you bind commandTemplate to `this` (in the constructor) so that you can use the same syntax as the other calls?

CalebJohn replied:
The reason for this syntax is because we need to pass command.value to the function. I changed the syntax of the call to fn, let me know what you think.

laurent22 commented on ElectronClient/app/gui/NoteText.jsx:
```diff
@@ -1835,6 +1855,12 @@ class NoteTextComponent extends React.Component {
 
 		const titleBarDate = <span style={Object.assign({}, theme.textStyle, {color: theme.colorFaded})}>{time.formatMsToLocal(note.user_updated_time)}</span>
 
+		const templatePrompt = <span className={body === '' ? 'fade_in' : 'fade_out'} style={templateStyle}>{_("Nothing here... try inserting a ")}
```
Maybe just `const templatePrompt = !this.props.templates.length ? null : ....` as that will remove a few conditionals below.

CalebJohn replied:
Done.

laurent22 commented on ReactNativeClient/lib/TemplateUtils.js:
```diff
@@ -0,0 +1,58 @@
+
+		if (await fs.pathExists(filePath)) {
+			try {
+				filenames = fs.readdirSync(filePath);
```
Please use shim.fsDriver().readDirStats (see fs-driver-node.js). In general async functions should be used to avoid blocking io, and shim.fsDriver should be used so that we can potentially use the same code in mobile too. Maybe not relevant in this particular case, but I prefer everything to go through fsDriver in general.

CalebJohn replied:
Thanks for the info, I've updated all uses of fs to shim.fsDriver().

laurent22 commented on ReactNativeClient/lib/TemplateUtils.js:
```diff
@@ -0,0 +1,58 @@
+		let templates = [];
+		let filenames = [];
+
+		if (await fs.pathExists(filePath)) {
```
shim.fsDriver().exists

laurent22 commented on ReactNativeClient/lib/TemplateUtils.js:
```diff
@@ -0,0 +1,58 @@
+			filenames.forEach(filename => {
+				if (filename.endsWith('.md')) {
+					try {
+						let fileString = fs.readFileSync(filePath + '/' + filename, 'utf-8');
```
`shim.fsDriver().readFile`

laurent22 commented on ReactNativeClient/lib/TemplateUtils.js:
```diff
@@ -0,0 +1,58 @@
+						let msg = error.message ? error.message : '';
+						msg = 'Could not load template ' + filename + '\n' + msg;
+						error.message = msg;
+						throw error;
```
Nice to wrap the error 👍

laurent22 commented:
I've left a few comments on the code. Overall there's nothing major to change I think, mostly the file system calls that should go through shim.fsDriver()

One issue I've noticed is that when I add my first template, I cannot use it right away without restarting the app. If I add a second template, it doesn't show up in the list of templates either.

I'm not sure what would be the solution here: either we watch the templates dir and update templates dynamically; or we provide an menu item to refresh the list of templates.

CalebJohn commented:
Thanks @laurent22 I've updated with your suggestions!

I think refresh button is the way to go, mostly because I don't really know how to set up a watcher, and ideally this system is just a stop gap until we have a more robust system of syncing templates and styles.

laurent22 commented on ElectronClient/app/gui/NoteText.jsx:
```diff
@@ -1027,7 +1027,7 @@ class NoteTextComponent extends React.Component {
 			} else if (command.name === 'textCode') {
 				fn = this.commandTextCode;
 			} else if (command.name === 'insertTemplate') {
-				fn = () => { return this.commandTemplate(command.value); };
```
No sorry you're right, it should be like you've done initially and it's also how it's done for focusElement below for instance. As long as the function is correctly bound (which it was in your code) it's fine. Sorry about that, it's a bit tricky to make sense of the diff sometime.

CalebJohn replied:
No worries, I thought this might happen, it's just worthwhile to have the code in place.

laurent22 commented:
Thanks @CalebJohn, we're nearly there I think. Just a few things remaining:
- One comment on the code, which was my fault. If you could change it as it was before that would be great
- Please also add `await` to the calls to refreshMenu() and updateMenu(). If they are called from non-async functions don't worry about it.
- I forgot about the template prompt on new note. I'm thinking it might be a bit too much because the common case is to create an empty note, and it also looks like this text is already in the note (while it's of course just a placeholder). So I'd prefer if it was not there by default. It's up to you, either make it optional (an option under the Note section for instance), or remove it, whichever you prefer. Also it can't be translated as it is, so if you keep it leave non-translated for now.

CalebJohn commented:
All done @laurent22 initially I tried to change the new note prompt to make it fit, but after a few minutes it became clear that you're right, it's not necessary, I've chosen to remove it rather than have it as a toggle.

laurent22 commented:
It's all good now, many thanks for your work @CalebJohn, much appreciated!

menghuu commented:
the template should also get the filename as the title
